roxie/roxiemem/roxiemem.cpp

```diff
@@ -5283,72 +5322,81 @@ void * CChunkedHeap::doAllocateRow(unsigned allocatorId, unsigned maxSpillCost)
                 }
             }
-        }
 
-        //NB: At this point activeHeaplet = NULL;
-        try
-        {
-            rowManager->checkLimit(1, maxSpillCost);
-        }
-        catch (IException * e)
-        {
-            //No pages left, but freeing the buffers may have freed some records from the current heap
-            if (!mayHaveEmptySpace())
-                throw;
-            logctx.CTXLOG("Checking for space in existing heap following callback");
-            e->Release();
-            continue;
-        }
+            //NB: At this point activeHeaplet = NULL;
+            //First perform a quick check - which is safe to execute within the critical section.
+            if (!rowManager->quickCheckLimit(1))
+            {
+                //This code needs to be outside the critical section, otherwise it may deadlock with another thread trying
+                //to dispose of any free pages.
+                block.leave();
 
-        donorHeaplet = allocateHeaplet();
-        if (donorHeaplet)
-            break;
-        rowManager->restoreLimit(1);
+                try
+                {
+                    rowManager->checkLimit(1, maxSpillCost);
+                }
+                catch (IException * e)
+                {
+                    //No pages left, but freeing the buffers may have freed some records from the current heap
+                    if (!mayHaveEmptySpace())
+                        throw;
+                    logctx.CTXLOG("Checking for space in existing heap following callback");
+                    e->Release();
+                    continue;
+                }
+            }
 
-        //Could check if activeHeaplet was now set (and therefore allocated by another thread), and if so restart
-        //the function, but grabbing the spin lock would be inefficient.
-        if (!rowManager->releaseCallbackMemory(maxSpillCost, true))
-            throwHeapExhausted(allocatorId, 1);
-    }
+            donorHeaplet = allocateHeaplet();
+            if (donorHeaplet)
+                break;
+            rowManager->restoreLimit(1);
 
-    if (memTraceLevel >= 5 || (memTraceLevel >= 3 && chunkSize > 32000))
-        logctx.CTXLOG("RoxieMemMgr: CChunkingRowManager::allocate(size %u) allocated new FixedSizeHeaplet size %u - addr=%p pageLimit=%u peakPages=%u rowMgr=%p",
-                chunkSize, chunkSize, donorHeaplet, rowManager->getPageLimit(), rowManager->peakPages, this);
+            //Could check if activeHeaplet was now set (and therefore allocated by another thread), and if so restart
+            //the function, but grabbing the spin lock would be inefficient.
+            if (!rowManager->releaseCallbackMemory(maxSpillCost, true))
+                throwHeapExhausted(allocatorId, 1);
+        }
 
-    {
-        NonReentrantSpinBlock b(heapletLock);
-        insertHeaplet(donorHeaplet);
+        if (memTraceLevel >= 3 && (memTraceLevel >= 5 || chunkSize > 32000))
```

Comment:
Actually I think that was deliberate.  The new code will short-circuit if traceLevel < 3 (the normal situation).  Previously it would have tested both or conditions.